workflow/docs/design.md

```diff
@@ -0,0 +1,406 @@
+# Internal Design of Chaos Mesh Workflow
+
+<!-- TOC -->
+
+- [Internal Design of Chaos Mesh Workflow](#internal-design-of-chaos-mesh-workflow)
+  - [Overview](#overview)
+  - [Core Concepts](#core-concepts)
+  - [Workflow Engine](#workflow-engine)
+    - [Trigger](#trigger)
+    - [Scheduler](#scheduler)
+    - [Actor and Playground](#actor-and-playground)
+    - [WorkflowManager: rest glue codes](#workflowmanager-rest-glue-codes)
+    - [States for Node](#states-for-node)
+    - [Node StateMachine](#node-statemachine)
+      - [Serial](#serial)
+      - [Suspend](#suspend)
+      - [Chaos](#chaos)
+    - [States for Workflow](#states-for-workflow)
+    - [Event](#event)
+    - [Others](#others)
+  - [Implement for Various Template](#implement-for-various-template)
+    - [Suspend](#suspend-1)
+    - [Parallel/Serial](#parallelserial)
+    - [Chaos](#chaos-1)
+    - [Task](#task)
+      - [Downward API](#downward-api)
+  - [Performance limitation](#performance-limitation)
+  - [API](#api)
+    - [Kubernetes CRD](#kubernetes-crd)
+  - [Schema-less Serialization](#schema-less-serialization)
+    - [Raw](#raw)
+    - [Unstructured](#unstructured)
+    - [v1.Json](#v1json)
+    - [mapstructure](#mapstructure)
+  - [Observability](#observability)
+    - [Logs](#logs)
+    - [Metrics](#metrics)
+    - [Tracing](#tracing)
+  - [Compare with v1.0.2](#compare-with-v102)
+    - [Breaking Changes](#breaking-changes)
+  - [Unresolved Problems](#unresolved-problems)
+
+<!-- /TOC -->
+
+> WIP: This document is still under development, everything is not stable, might change frequently.
+
+## Overview
+
+As the RFC of Chaos Mesh Workflow is stable, we create this document as the reference for implementation.
+
+We will not define the certain struct here, but we present pseudo-code for interface and main logic.
+
+An instance of Chaos Mesh Workflow is basically like a tree:
+
+- each action is described by a node;
+- a root node represents for entry-point;
+- nodes with certain types could have children nodes;
+- depends on the type of this node, children nodes execute serial or parallel;
+- node with certain type could select which children nodes to run(conditional branch);
+
+So Chaos Mesh Workflow is mostly like:
+
+- Argo
+
+It doesn't like:
+
+- BPMN
+
+> Do we really need a tree? How about a pipeline?
+
+## Core Concepts
+
+**Workflow**: a resource that defines the orchestration of chaos experiments.
+
+**Template**: the node in **Workflow** "tree", represents for operation. There are various operations: like "create chaos experiments", "wait for 5 minutes", "create a serial job with children templates", etc.
+
+**Node**: a running/completed step within a workflow, it represents the status; We could think of it as an instance rendered from a **Template**.
```

Comment:
It is said that `Template` is the node in **Workflow** "tree", so what is the relationship between `Template` and `Node`?

Reply:
The relationship between `Template` and `Node` just like `Class` and `Object` in OOP.